### ISSUE TITLE
sentry-cli to upload to external symbol store

### ISSUE THREAD

**Issue by kristjanvalur** [getsentry/sentry-cli#947]
So, Yesterday I started to work with symbol files in my project and _immediately_ hit the 2GB dif file barrier.  It was suggested to me via some of the issues here that the only workaround was to set up an external symbol store.
To even access this feature, I enabled the business plan for my org, but I was dismayed to find that the only thin that it did was to enable a read-only access to some external bucket.
No convenient way to actually add symbols.

sentry-cli is very convenient.  You point it at a folder and it trawls it for appropriate binaries, queries the upstream if they exist already, and only uploads those that matter.

Using an external symbol repo, to bypass sentry storage limits, removes this convenience completely.
Would it

-  Would it be hard to simply give sentry read/write access to an external bucket, like s3?  Do the file offloading automatically?
- Alternatively, allow the sentry-cli to talk directly to an external store, such as s3 or http, and upload directly, probably using the "unified" directory layout?

Basically, I just want a way to bypass the 2GB limit in a convenient way.  Also, being able to periodically prune my external store is convenient, but I can write my own tool to do that.

**Comment by kamilogorek** [comment]
> Alternatively, allow the sentry-cli to talk directly to an external store, such as s3 or http, and upload directly, probably using the "unified" directory layout?

We do have a tool for exactly this usecase which you should be able to solve this specific scenario: https://github.com/getsentry/symbolicator/tree/master/symsorter

**Comment by kristjanvalur** [comment]
Researching further, I did find the https://github.com/getsentry/coredump-uploader
Perhaps this is a better alternative when working with linux?  It doesn't require uploading any symbols at all (apparently, linux core files are not directly supported, only _minidump_ files.  the documentation is confusing on this topic)

**Comment by flub** [comment]
(oops, found this in an open browser tab with a half-finished response...)

As @kamilogorek already said there's a, fairly rough, "symsorter" tool which helps you create the required directory structures.  From there it is probably reasonable enough to script things to upload to S3 automatically.

However as said the tool is fairly rough, we don't even make releases for it currently so you need to build it yourself.  Feel free to file bugs or features requests at https://github.com/getsentry/symbolicator if things could be better to support your workflow.

**Comment by jan-auer** [comment]
@kristjanvalur The coredump uploader would require debug symbols for all libraries and your own application on the crashing machine to show full stack traces. Since it uses `gdb` under the hood, it will try to symbolicate as much as possible before sending data to Sentry. 

We developed the coredump uploader as an experimental alternative where using our SDKs and/or minidumps is not possible or where existing coredump infrastructure is in place. We do not maintain and support it to the same level as our SDKs. For production use, I recommend setting up a symbol server, instead.

**Comment by kristjanvalur** [comment]
As I see it, there are two alternatives, then:
1 - Upload symbols during the build process.  Use an external symbol store
because of the 2GB limit.  Run a special black-box server alongside my
server process that creates mysterious minidump files out of the `core`
files.  Build and maintain google-crashpad.
2 - Keep symbols with my server.  When it crashes, coredump-uploader
analyzes it and uploads symbols.

From my point of view, number two is a much simpler proposition.  It avoids
the entire symbol file build entirely, and uses python.  Installing gdb and
binutils is super simple and keeping the symbols as part of the binaries of
the server that I am running, is simple.

Server deployment is different in this aspect than client tooling.  I am
using this for Unreal Server, by the way, and the UE4 doesn't do any sort
of automatic crash dump uploading when it is a linux build.

I'll continue trying to get this to work, and report back here to you, but
so far it looks to be much simpler than the alternative.

mán., 19. apr. 2021 kl. 08:22 skrifaði Jan Michael Auer <
***@***.***>:

**Comment by jan-auer** [comment]
That's right, provided deploying debugging information to your Linux servers is an option then you can go with the coredump uploader.

If I'm not mistaken, UE4's crash reporter still does not generate crash reports on Linux. Let me do some more research on whether there are any news in recent engine versions. We would have an [integration for the built-in crash reporter](https://docs.sentry.io/platforms/native/guides/ue4/). In that case, it would still be required to upload files to a symbol server, however.

As for _alternative 1_, you could also run our [Native SDK](https://github.com/getsentry/sentry-native) which will take care of capturing crashes and uploading them to Sentry, even allowing for capturing custom context and breadcrumbs. We have an "inproc" backend that can also symbolicate using debug info on your server, just like GDB.

**Comment by kristjanvalur** [comment]
Right, the crash reporter is woefully undocumented.  It is not even packaged for linux, I suspect that it, in fact, isn't functional on linux.  Not from what I could glean from unreal .cs code, anyway.
I _am_ using that setup (and those instructions) for client builds, however, where we are using windows, and will be using symbol uploads.  That is more familiar territory.

I will also look at the Native SDK as suggested.

Meanwhile, you may expect PRs for the crashdump-uploader, we're giving it a shakedown :)

**Comment by kristjanvalur** [comment]
Well, got it working, needed tweaking to work with py3, and minor fixes.  But in truth, a bit underwhelming.
I'm now looking at the native SDK.  I'm wondering at the feasability of simply making it an UE4 plugin, with precompiled libraries.  The MS CRT thing might be problematic...

I'm also a bit confused about the backend.  Is there any reason to not use the inproc backend?  Also, I wonder if using this plugin, even on a client, might be preferable to using epic's own crash reporting client....

**Comment by kristjanvalur** [comment]
Ok, Just a few more questions before i embark upon testing this...
1. Will the out of process backend be automatically launched, or are there steps to be taken?
2. Is there a way for a process runner to make sure that any uploads have finished?  I'm thinking about servers running in a container here. My container_entrypoint mustn't exit until the backend has finished uploading.

**Comment by flub** [comment]
1. If the sentry SDK is configured with a DSN it will automatically launch the out of process backend without further action, if it can find it (normally same directory as binary or configurable via options).
2. I believe once the crashpad_handler binary is terminated you can be sure the upload is finished, but I'm not 100% sure on this.  So if you ensure you do not terminate the container before that process has finished then it should be fine.

**Comment by jan-auer** [comment]
The `crashpad_handler` process completely detaches from it's launcher using the double-fork pattern. This allows it to continue running even if the parent (which it monitors) crashes. It also retries failed uploads on network downtime or submission failure.

Only other thing I can think of: You will still have to ensure the handler can be launched, so it has to be packaged with the game and you need to know the path during runtime.

**Comment by kristjanvalur** [comment]
Right, it forks.  That's fine, then.
the remaining bit is then now to wait for it to exit.  the shell must be
sure that both the game process, and the crashpad_handler process have
exited, before
it exits itself, otherwise the crashpad_handler process will be prematurely
terminated as the container stops.
I'm sure I"ll figure something out with bash...
```
# start game
run_game_server
code=$?
#wait for crashpad to exit, if it needs to upload
while ps -ef | grep crashpad ; do
  sleep 0.1
#exit container
exit $code
```

þri., 20. apr. 2021 kl. 14:57 skrifaði Jan Michael Auer <
***@***.***>:

**Comment by jan-auer** [comment]
You could check if `crashpad_handler` turns up as a child of your shell. It probably doesn't, but if it does you could just run `wait $(pgrep crashpad_handler | head -1)` after your process exits. Otherwise, running a loop around `[ -e "/proc/$(pgrep crashpad_handler | head -1)" ]` should do the trick.

For the record: The more complicated version would be to start the handler first and then attach our SDK to the running instance. The crashpad client would support attaching to a running handler, but we haven't exposed any of that yet in our API.

**Comment by kristjanvalur** [comment]
> Most definitely, and to say the least, we'd be interested in officially supporting and maintaining such a plugin in the long run. Recently, it has not been on our immediate roadmap, but we can definitely provide assistance if you're interested in building it out.
> 
Here it is! https://github.com/mainframeindustries/UE4SentryClientPlugin
This is initial release, with blueprint support for most of the api, plus crash handling.  Build for Win64 and Linux initially.
Please see the specific cmake hoops I had to jump throught.
For windows crash handling to work, it needs this PR: https://github.com/EpicGames/UnrealEngine/pull/7976

**Comment by kristjanvalur** [comment]
> (oops, found this in an open browser tab with a half-finished response...)
> 
> As @kamilogorek already said there's a, fairly rough, "symsorter" tool which helps you create the required directory structures. From there it is probably reasonable enough to script things to upload to S3 automatically.
> 
> However as said the tool is fairly rough, we don't even make releases for it currently so you need to build it yourself. Feel free to file bugs or features requests at https://github.com/getsentry/symbolicator if things could be better to support your workflow.

So, I built symsorter and tried it out and it is not very pleasant.

First off, it needs some `bundle-id` which I have no idea what is, or why it is required.  Also a Prefix.  `sentry-cli` doesn't need either.

Pointing it at my build folder, it will process all .o files that it finds, and sort them into some 00000 guid....   Why is it messing with my object files?
I must confess I have no idea what is supposed to happen.
But with sentry-cli, you point it at a folder, it finds all the symbol files automatically and uploads only those.  Further, it will know which files it has previously uploaded (based on a hash, presumably) and not upload them.
I wish I could use something as simple as that, and upload my symbols to sentry, having _large_ files automatically offloaded ot my own storate, for instance.
I would also be happy if I could point sentry-cli at my s3 bucket and say, "upload to there".
Mostly I just want to overcome the 2 GB limit.

**Comment by kristjanvalur** [comment]
Well, I found that if I point it at a folder that has none of the intermediate .o files, then it behaves much better.  It would be very useful to be able to ask symsorter to ignore such files.  

Still, confused about the `prefix` and the `bundle-id` and how that will affect things.  Will continue now with tryint to upload my output to s3 and hook that up to my sentry project...

**Comment by kristjanvalur** [comment]
Getting a lot of the following errors when symsorint UE4 engine symbols (Win64, debug)
```
SwarmAgent.exe (exe, x86) -> d:\scratch\output2\Win64\a6/153c0e-3e1a-4667-83f9-621f6a4fc7c3/executable
thread '<unnamed>' panicked at 'thread 'called `Result::unwrap()` on an `Err` value: ParseDebugIdError<unnamed>', ' panicked at 'symsorter\src\app.rscalled `Result::unwrap()` on an `Err` value: ParseDebugIdError:', 139symsorter\src\app.rs:thread ':64139<unnamed>
' panicked at ':called `Result::unwrap()` on an `Err` value: ParseDebugIdErrorthread '64thread '', <unnamed>
<unnamed>symsorter\src\app.rs' panicked at '' panicked at 'thread ':called `Result::unwrap()` on an `Err` value: ParseDebugIdErrorcalled `Result::unwrap()` on an `Err` value: ParseDebugIdError<unnamed>139' panicked at '', thread '', :called `Result::unwrap()` on an `Err` value: ParseDebugIdErrorthread 'symsorter\src\app.rs<unnamed>symsorter\src\app.rs64', <unnamed>thread ':<unnamed>:
symsorter\src\app.rs' panicked at '' panicked at '139' panicked at '139:called `Result::unwrap()` on an `Err` value: ParseDebugIdErrorcalled `Result::unwrap()` on an `Err` value: ParseDebugIdError:called `Result::unwrap()` on an `Err` value: ParseDebugIdError:139', ', 64symsorter\src\app.rs64:thread '64',
:
```

**Comment by kristjanvalur** [comment]
Oh dear!  Symsorter will, despite the documentation, not create a "unified" folder structure.
This is what Sentry tells me, when it cannot find my symbol:
```
not found: s3://paxdei-symbols-mainframe-zone/5e/3aa57163c94b9a89e0e1d0b2ab6c081/debuginfo
```
However, the folder on disk is:
`D:\scratch\output2\Win64\5e\3aa571-63c9-4b9a-89e0-e1d0b2ab6c08`
which I subsequently uploaded.  Note the hyphens, and also the missing `1` at the end....    I'll create an issue there to report this.

**Comment by kristjanvalur** [comment]
@jan-auer did you get a chance to look at the Sentry UE4 plugin yet, as mentioned above?

**Comment by jan-auer** [comment]
Sorry, @kristjanvalur, I'm a little backed up lately. Just checking in that we're not forgetting about this.
